benches/setup.rs

```diff
@@ -1,116 +1,93 @@
 use scryer_prolog::machine::{
     parsed_results::{QueryMatch, QueryResolution, Value},
     Machine,
 };
 
-pub fn benches() -> Benches {
-    Benches::new(&[
+pub fn prolog_benches() -> BTreeMap<&'static str, PrologBenchmark> {
+    [
+        (
+            "count_edges",                       // name of the benchmark
+            "benches/edges.pl",                  // name of the prolog module file to load
+            "independent_set_count(aa, Count).", // query to benchmark in the context of the loaded module
+            Strategy::Reuse,
+            btreemap! { "Count" => Value::try_from("211954906".to_string()).unwrap(), }, // list of expected bindings
+        ),
         (
-            "benches/edges.pl", // name of the prolog module file to load
-            &[
-                (
-                    "count_edges_short",                 // name of the benchmark
-                    "independent_set_count(ky, Count).", // query to benchmark in the context of the loaded module
-                    btreemap! { "Count".to_string() => Value::try_from("2869176".to_string()).unwrap() }, // List of expected bindings
-                ),
-                (
-                    "count_edges", // multiple benchmark queries can be defined per module
-                    "independent_set_count(aa, Count).", // consider making the query adjustable to tune the runtime
-                    btreemap! { "Count".to_string() => Value::try_from("211954906".to_string()).unwrap(), },
-                ),
-            ],
+            "count_edges_short",
+            "benches/edges.pl", // use the same file in multiple benchmarks
+            "independent_set_count(ky, Count).", // consider making the query adjustable to tune the run time to ~0.1s
+            Strategy::Reuse,
+            btreemap! { "Count" => Value::try_from("2869176".to_string()).unwrap() },
         ),
         (
+            "numlist_short",
             "benches/numlist.pl",
-            &[(
-                "numlist_short",
-                "run_numlist(1000000, Head).",
-                btreemap! { "Head".to_string() => Value::try_from("1".to_string()).unwrap()},
-            )],
+            "run_numlist(1000000, Head).",
+            Strategy::Reuse,
+            btreemap! { "Head" => Value::try_from("1".to_string()).unwrap()},
+        ),
+        (
+            "csv_codename",
+            "benches/csv.pl",
+            "get_codename(\"0020\",Name).",
+            Strategy::Reuse,
+            btreemap! { "Name" => Value::try_from("SPACE".to_string()).unwrap()},
         ),
-    ])
+    ]
+    .map(|b| {
+        (
+            b.0,
+            PrologBenchmark {
+                name: b.0,
+                filename: b.1,
+                query: b.2,
+                strategy: b.3,
+                bindings: b.4,
+            },
+        )
+    })
+    .into()
 }
 
-pub struct Benches {
-    machines: Vec<Machine>,
-    runs: BTreeMap<String, Run>,
+pub enum Strategy {
```

Comment:
If I understand correctly, Strategy does nothing at this moment right?

Reply:
Strategy is used in `run_criterion.rs`. When a test is labeled as `Fresh` then it creates batches of fresh machines each used exactly once in order to support benchmarks that change the machine state (say, benchmarking assertz).